### PR TITLE
Implement `try_fold` for `Zip` iterator

### PR DESCRIPTION
**Pull request by JohnTitor**
This is a rework of #60102.
Implements `try_fold` for `Zip` iterator.

r? @sfackler

### CODE REVIEW

File: src/libcore/iter/adapters/zip.rs [hunk 1]
```diff
@@ -51,6 +53,14 @@ impl<A, B> Iterator for Zip<A, B> where A: Iterator, B: Iterator
     fn nth(&mut self, n: usize) -> Option<Self::Item> {
         ZipImpl::nth(self, n)
     }
+
+    #[inline]
+    fn try_fold<Acc, F, R>(&mut self, init: Acc, f: F) -> R
+        where F: FnMut(Acc, Self::Item) -> R,
```
Comment (lzutao) [inline src/libcore/iter/adapters/zip.rs]: fmt?

File: src/libcore/iter/adapters/zip.rs [hunk 2]
```diff
@@ -144,6 +153,20 @@ impl<A, B> ZipImpl<A, B> for Zip<A, B>
 
         (lower, upper)
     }
+
+    #[inline]
+    default fn try_fold<Acc, F, R>(&mut self, init: Acc, mut fold: F) -> R
+        where F: FnMut(Acc, Self::Item) -> R,
```
Comment (lzutao) [inline src/libcore/iter/adapters/zip.rs]: fmt?

File: src/libcore/iter/adapters/zip.rs [hunk 3]
```diff
@@ -75,6 +85,9 @@ trait ZipImpl<A, B> {
     fn next_back(&mut self) -> Option<Self::Item>
         where A: DoubleEndedIterator + ExactSizeIterator,
               B: DoubleEndedIterator + ExactSizeIterator;
+    fn try_fold<Acc, F, R>(&mut self, init: Acc, f: F) -> R
+        where F: FnMut(Acc, Self::Item) -> R,
```
Comment (lzutao) [inline src/libcore/iter/adapters/zip.rs]: fmt?

**Comment by czipperz** [comment]
Do we have perf that shows a gain here?  It would be nice to have something concrete.

**Comment by czipperz** [comment]
Please add some tests to this behavior to ensure it works correctly.

**Comment by JohnTitor** [comment]
Ya, but before for them, we should pass the test failed in CI. I don't come up with the solution...

**Comment by czipperz** [comment]
You could try copy/pasting the default implementation of `try_fold` for iterators and seeing if that passes.

**Comment by sinkuu** [comment]
[`ZipImpl` specialization for `TrustedRandomAccess`](https://doc.rust-lang.org/1.36.0/src/core/iter/adapters/zip.rs.html#150) does not advance underlying iterators, which makes the direct use of the underlying iterator's methods invalid within such `ZipImpl`.

```rust
let mut it = /* `zip` of TrustedRandomAccess iterators */;

it.next(); // only increases `index` field of `Zip`
it.try_fold( ... ); // this PR: starts iteration from the begenning
```

This use-case [seemingly appears in the failing test.](https://github.com/rust-lang/rust/blob/0324a2b309cd66cb7bd4a156bd0b84cb136e254f/src/test/run-pass/impl-trait/example-calendar.rs#L661-L667)

Given that `ZipImpl::next` specialized for `TrustedRandomAccess` iterators is (or should be) fast, I think you can just copy [`Iterator::try_fold`'s default implementation](https://doc.rust-lang.org/1.36.0/src/core/iter/traits/iterator.rs.html#1567-1575) to specialized `ZipImpl`.

**Comment by timvermeulen** [comment]
>Given that `ZipImpl::next` specialized for `TrustedRandomAccess` iterators is (or should be) fast, I think you can just copy `Iterator::try_fold`'s default implementation to specialized `ZipImpl`.

What's the point of doing that? The only reason for implementing `Zip::try_fold` in the first place is that its implementation can call `try_fold` of one of the underlying iterators (which for some iterators is faster than repeated `next` calls). Copying the default implementation is by definition not going to change anything.

**Comment by sinkuu** [comment]
@timvermeulen Of course, the current state of this PR does not make sense. At the time I made that comment, this PR contained `try_fold` implementation using underlying iterator's `try_fold`, which could be possibly faster than default implementation, but doesn't work for `TrustedRandomAccess` iterators. I meant keeping that implementation for `A, B: !TrustedRandomAccess` case, and use the default implementation in `A, B: TrustedRandomAccess`:

```rust
impl<A, B> ZipImpl<A, B> for Zip<A, B> {
    fn try_fold<Acc, F, R>(&mut self, init: Acc, mut f: F) -> R where
        Self: Sized, F: FnMut(Acc, Self::Item) -> R, R: Try<Ok=Acc> {
        self.a.try_fold(...)
    }
}

impl<A, B> ZipImpl<A, B> for Zip<A, B>
    where A: TrustedRandomAccess, B: TrustedRandomAccess
{
    fn try_fold<Acc, F, R>(&mut self, init: Acc, mut f: F) -> R where
        Self: Sized, F: FnMut(Acc, Self::Item) -> R, R: Try<Ok=Acc> {
        // Use `Iterator::try_fold` implementation
    }
}
````

**Comment by JohnTitor** [comment]
I can't come up with suitable tests for this - so I'm going to close this.

**Pull request closed without merging**